apps/ejabberd/src/jlib.erl

```diff
@@ -268,67 +268,69 @@ iq_query_info(El) ->
-                                'invalid' | 'not_iq' | 'reply' | ejabberd:iq().
+  'invalid' | 'not_iq' | 'reply' | ejabberd:iq().
 iq_info_internal(#xmlel{name = Name, attrs = Attrs,
-                        children = Els}, Filter) when Name == <<"iq">> ->
-    %% Filter is either request or any.  If it is request, any replies
-    %% are converted to the atom reply.
-    ID = xml:get_attr_s(<<"id">>, Attrs),
-    Type = xml:get_attr_s(<<"type">>, Attrs),
-    Lang = xml:get_attr_s(<<"xml:lang">>, Attrs),
-    {Type1, Class} = case Type of
-                         <<"set">> -> {set, request};
-                         <<"get">> -> {get, request};
-                         <<"result">> -> {result, reply};
-                         <<"error">> -> {error, reply};
-                         _ -> {invalid, invalid}
-                     end,
-    if
-        Type1 == invalid ->
-            invalid;
-        Class == request; Filter == any ->
-            %% The iq record is a bit strange.  The sub_el field is an
-            %% XML tuple for requests, but a list of XML tuples for
-            %% responses.
-            FilteredEls = xml:remove_cdata(Els),
-            {XMLNS, SubEl} =
-                case {Class, FilteredEls} of
-                    {request, [#xmlel{attrs = Attrs2}]} ->
-                        {xml:get_attr_s(<<"xmlns">>, Attrs2),
-                         hd(FilteredEls)};
-                    {reply, _} ->
-                        %% Find the namespace of the first non-error
-                        %% element, if there is one.
-                        NonErrorEls = [El ||
-                                          #xmlel{name = SubName} = El
-                                              <- FilteredEls,
-                                          SubName /= <<"error">>],
-                        {case NonErrorEls of
-                             [NonErrorEl] ->
-                                 xml:get_tag_attr_s(<<"xmlns">>, NonErrorEl);
-                             _ ->
-                                 <<>>
-                         end,
-                         FilteredEls};
-                    _ ->
-                        {<<>>, []}
-                end,
-            if XMLNS == <<>>, Class == request ->
-                    invalid;
-               true ->
-                    #iq{id = ID,
-                        type = Type1,
-                        xmlns = XMLNS,
-                        lang = Lang,
-                        sub_el = SubEl}
-            end;
-        Class == reply, Filter /= any ->
-            reply
-    end;
+  children = Els}, Filter) when Name == <<"iq">> ->
+  %% Filter is either request or any.  If it is request, any replies
+  %% are converted to the atom reply.
+  ID = xml:get_attr_s(<<"id">>, Attrs),
+  Type = xml:get_attr_s(<<"type">>, Attrs),
+  Lang = xml:get_attr_s(<<"xml:lang">>, Attrs),
+  {Type1, Class} = case Type of
```

Comment:
I think extracting this case to a function may improve readability of this function a bit.